psi4/src/psi4/libmints/mcmurchiedavidson.cc

```diff
@@ -0,0 +1,122 @@
+/*
+ * @BEGIN LICENSE
+ *
+ * Psi4: an open-source quantum chemistry software package
+ *
+ * Copyright (c) 2007-2022 The Psi4 Developers.
+ *
+ * The copyrights for code used from other parties are included in
+ * the corresponding files.
+ *
+ * This file is part of Psi4.
+ *
+ * Psi4 is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, version 3.
+ *
+ * Psi4 is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License along
+ * with Psi4; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * @END LICENSE
+ */
+#include "mcmurchiedavidson.h"
+
+namespace mdintegrals {
+
+inline int cart_dim(int L) { return (L + 1) * (L + 2) / 2; }
+
+inline int cumulative_cart_dim(int L) { return ((L + 1) * (L + 2) * (L + 3)) / 6; }
+
+std::vector<std::array<int, 4>> generate_am_components_cca(int am) {
+    std::vector<std::array<int, 4>> ret(cart_dim(am));
+    int index = 0;
+    for (int l = am; l > -1; --l) {
+        for (int n = 0; n < am - l + 1; ++n) {
+            int m = am - l - n;
+            ret[index] = {{l, m, n, index}};
+            index++;
+        }
+    }
+    return ret;
+}
+
+void fill_E_matrix(int maxam1, int maxam2, const Point& P, const Point& A, const Point& B, double a, double b,
+                   std::vector<double>& Ex, std::vector<double>& Ey, std::vector<double>& Ez) {
+    // computes the Hermite Gaussian expansion coefficients E_t^{ij} (eq 9.5.1)
+    // equation numbers from Molecular Electronic-Structure Theory (10.1002/9781119019572)
```

Comment:
This code is correct, and it's not likely to be rate-limiting for the potential ints.  If benchmarks show that it is taking time, we can consider unrolling the loops to remove the if statements, [like we do for the OS recursions](https://github.com/psi4/psi4/blob/9f7b865b0cecf694e83e04637ab6d78b4480e43f/psi4/src/psi4/libmints/osrecur.cc#L1344-L1392).  That would remove branches from the inner loops, and I think would speed it up.  It's not important to change unless we see this routine taking a significant amount of time.  I see you already partially do this for the t loops.

Reply:
I've done some profiling for my standalone C++ code and filling the E-matrix was not a rate-limiting step at all, so I think it should be fine for now 👍